diff --git a/gpu/command_buffer/client/client_transfer_cache.cc b/gpu/command_buffer/client/client_transfer_cache.cc
index 00d7a7058340d..3dc5d937f7e25 100644
--- a/gpu/command_buffer/client/client_transfer_cache.cc
+++ b/gpu/command_buffer/client/client_transfer_cache.cc
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 #include "gpu/command_buffer/client/client_transfer_cache.h"
+#include "gpu/command_buffer/client/cmd_buffer_helper.h"
 
 namespace gpu {
 
@@ -51,12 +52,16 @@ void ClientTransferCache::UnmapAndCreateEntry(uint32_t type, uint32_t id) {
 
   if (mapped_ptr_) {
     DCHECK(!transfer_buffer_ptr_);
+    client_->cmd_buffer_helper()->SendBufferToWhist(CommandBufferType::TRANSFER_CACHE,
+                  mapped_ptr_->address(), mapped_ptr_->size());
     client_->IssueCreateTransferCacheEntry(
         type, id, handle.shm_id(), handle.byte_offset(), mapped_ptr_->shm_id(),
         mapped_ptr_->offset(), mapped_ptr_->size());
     mapped_ptr_ = absl::nullopt;
   } else {
     DCHECK(!mapped_ptr_);
+    client_->cmd_buffer_helper()->SendBufferToWhist(CommandBufferType::TRANSFER_CACHE,
+                  transfer_buffer_ptr_->address(), transfer_buffer_ptr_->size());
     client_->IssueCreateTransferCacheEntry(
         type, id, handle.shm_id(), handle.byte_offset(),
         transfer_buffer_ptr_->shm_id(), transfer_buffer_ptr_->offset(),
